### DISCUSSION TITLE
Why Internal/Hidden Functions?

### DISCUSSION THREAD

**Discussion by mscirri** [dataplat/dbatools#7403] [Q&A] [answered]
The source code for dbatools is available for anyone to download and use, but not all functions are exported.   What is the point of having internal/hidden functions in the case where source code is available to everyone?   

BACKGROUND:  I have a project where Find-DbaLoginInGroup got me like 95% of the way there but I needed it to do one more thing, so I thought I would just copy the function code into my own function, add my customization, and as long as I had dbatools installed, it would work.   (I did add a suggestion for this specific feature here:  https://github.com/sqlcollaborative/dbatools/discussions/7402)

**Comment by niphlod** [comment]
if you import the psm1 file you get pretty much everything. That being said, all code gets used... "hidden" pieces are usually reused withing the public functions, which are the only ones with tested/approved/granted "entry-points"/"interfaces" dbatools exposes.

**Reply by mscirri** [reply]
Thank you for engaging.  Yeah, unfortunately, the psm1 file is missing a bunch of functions that are used with the original source code.   

I guess I just don't understand, in the world of open-source PowerShell (I am an experienced C# developer but rarely use PowerShell),  why not just export everything to make it easy for folks to utilize and customize?   I mean, I could just compile my own version of dbatools with the source code plus my customizations, but for the small thing that I need that seems like overkill.   Maybe it is just the mindset of someone who isn't a full-time developer anymore.

**Reply by niphlod** [reply]
ahem, what "bunch" ?
Also, what do you expect to do when "adding your own version of Find-DbaLoginInGroup" ?

**Reply by mscirri** [reply]
Functions not exported:   

Stop-Function
Connect-SqlInstance
Select-DefaultView
Convert-DbaMessageException
Convert-DbaMessageTarget

That is as far as I got before I stopped copying source code into my script file and decided there had to be a better way.

**Reply by mscirri** [reply]
As far as what I am trying to do, I added this:  https://github.com/sqlcollaborative/dbatools/discussions/7402

**Reply by mscirri** [reply]
And taking a step back even further, here is what this mini project is supposed to accomplish:

I need a list of all of the AD users that have access to certain SQL Server environments and I need to know what access they have, whether at the server or database level.   This is for compliance and internal auditing purposes.    Most of our permissions are granted through AD Security groups, and some groups contain other groups a members.    We also have some permissions that are granted on a DB but without a SQL login.

I hope that provides some context.

**Reply by niphlod** [reply]
all of those are available if you import the psm1 file in "developer mode" (fastest way is making sure a .git folder exists in the directory where dbatools.psm1 and dbatools.psd1 are)

**Reply by mscirri** [reply]
OK. I have never done that before so I will have to take a moment look that up, but I will definitely give it a shot.

**Reply by mscirri** [reply]
It still does not seem to recognize the 2 main functions I have been having issues with. I tried just importing the PSM1 file as well as importing dbatools AND the PSM1 file.  Same result unfortunately.

===========================================

Stop-Function : The term 'Stop-Function' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Users\mscirri\OneDrive - Empire Innovation Group LLC\Documents\PS Scripts\SQL Server\Find-DbaLoginInGroupWithDBOnlyUsers.ps1:125 char:17
                 Stop-Function -Message "Error occurred while establis ...
                 ~~~~~~~~~~~~~
     CategoryInfo          : ObjectNotFound: (Stop-Function:String) [], CommandNotFoundException
     FullyQualifiedErrorId : CommandNotFoundException
 
Select-DefaultView : The term 'Select-DefaultView' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.   
At C:\Users\mscirri\OneDrive - Empire Innovation Group LLC\Documents\PS Scripts\SQL Server\Find-DbaLoginInGroupWithDBOnlyUsers.ps1:163 char:13
             Select-DefaultView -InputObject $res -Property SqlInstanc ...
             ~~~~~~~~~~~~~~~~~~
     CategoryInfo          : ObjectNotFound: (Select-DefaultView:String) [], CommandNotFoundException
     FullyQualifiedErrorId : CommandNotFoundException

**Comment by andreasjordan** [comment, marked as answer]
Parameters or functionality of internal commands could change without notice.
But there are indeed some internal function I would like to have exported so that I can use them myself.

**Reply by potatoqualitee** [reply]
Interesting! Which ones, @andreasjordan ?

**Reply by andreasjordan** [reply]
The last that I wanted to use was [Test-PendingReboot](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Test-PendingReboot.ps1) to test for pending reboots before starting Install-DbaInstance (and because Install-DbaInstance currently does not test for pending reboots because of open file renames).
I already implemented [Convert-ByteToHexString](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Convert-ByteToHexString.ps1) myself, so this and [Convert-HexStringToByte](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Convert-HexStringToByte.ps1) is a good candidate.
I once needed [Get-DbaHelp](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Get-DbaHelp.ps1) - and will need it again to get some better formatted documentation for a command.
And maybe [Invoke-Command2](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Invoke-Command2.ps1) and [Invoke-CommandWithFallback](https://github.com/sqlcollaborative/dbatools/blob/development/internal/functions/Invoke-CommandWithFallback.ps1) would be candidates also.

But exporting functions that don't start with "-Dba" is always a risk because of name collisions. Maybe change these names and then export. 

Not an easy topic...

**Reply by potatoqualitee** [reply]
Ah, those are really good, but admittedly out of scope, except perhaps Test-PendingReboot. I can see a lot of DBAs needing that one. 

You can create a helper module with the commands you like and release it to the Gallery. I tend to just copy the commands I like to my other modules and add them to my modules.
